server/aesdsocket.c

```diff
@@ -0,0 +1,303 @@
+                break;
+        }
+
+		// Send whatever is read to client
+		send(socketFd, buf, (ret/sizeof(char)), 0);
+	}
+}
+
+/*
+ * Main: Entry point for the program
+ */
+int main(int argc, char* argv[])
+{	
+	bool daemonMode = false;
+	int rc;
+	struct addrinfo hints;
+	struct sockaddr clientInfo;
+	socklen_t clientInfoLen = sizeof clientInfo;
+
+	// Check arguments
+	while((rc = getopt(argc, argv,"d")) != -1)
+    {
+        switch(rc)
+        {
+            case 'd':
+                daemonMode = true;
+                break;
+        }
+
+    }
+	
+	//Register for signals
+	signal(SIGINT, SignalHandler);
+	signal(SIGTERM, SignalHandler);
+	
+	// Set up syslog
+	openlog("aesdsocket", LOG_NDELAY, LOG_USER);
+	
+	// Create socket
+	serverSockfd = socket(PF_INET, SOCK_STREAM, 0); // IPv4, TCP socket
+	if(serverSockfd == -1)
+	{
+		perror("socket");
+		return -1;
+	}
+
+	// Set up socket to reuse the address
+	int a = 1;
+	rc = setsockopt(serverSockfd, SOL_SOCKET, SO_REUSEADDR, &a, sizeof(int));
+	if(rc == -1)
+	{
+		perror("setsockopt");
+		return -1;
+	}
+	
+	// Get addrinfo for socket
+	memset(&hints, 0, sizeof hints);
+	hints.ai_flags = AI_PASSIVE;
+	hints.ai_family = AF_INET;
+    	hints.ai_socktype = SOCK_STREAM;
+	
+	rc = getaddrinfo(NULL, PORT_NUM_STR, &hints, &serverInfo);
+	if(rc != 0)
+	{
+		printf("getaddrinfo: %s", gai_strerror(rc));
+		if(rc == EAI_SYSTEM)
+		{
+			perror("getaddrinfo");
+		}
+		close(serverSockfd);
+		return -1;
+	}
+	
+	
+	// Bind socket with the info
+	rc = bind(serverSockfd, serverInfo->ai_addr, sizeof(struct sockaddr));
+	if(rc == -1)
+	{
+		perror("bind");
+		freeaddrinfo(serverInfo);
+		close(serverSockfd);
+		return -1;
+	}
+
+	// Run in daemon mode based on flag
+	if(daemonMode)
+	{
+		rc = daemon(-1, -1); // Non-zero values to perform daemon operations
+		if(rc == -1)
+		{
+			perror("daemon");
+			freeaddrinfo(serverInfo);
+			close(serverSockfd);
+			return -1;
+		}
+	}
+		
+	// Listen for a connection
+	rc = listen(serverSockfd, MAX_PENDING_CONNECTIONS); // max number of pending connections: 3
+	if(rc == -1)
+	{
+		perror("listen");
+		freeaddrinfo(serverInfo);
+		close(serverSockfd);
+		return -1;
+	}
+	
+	while(!terminate)
+	{
+		// Accept the connection
+		acceptedSocketfd = accept(serverSockfd, &clientInfo, &clientInfoLen);
+		if(acceptedSocketfd == -1)
+		{
+			perror("accept");
+			freeaddrinfo(serverInfo);
+			close(serverSockfd);
+			return -1;
+		}
+		connectionInProgress = true;
+		char clientIpStr[INET6_ADDRSTRLEN];
+		inet_ntop(AF_INET, &clientInfo, clientIpStr, sizeof(clientIpStr));
+		syslog(LOG_INFO, "Accept connection from %s", clientIpStr); // Log client IP to syslog
+	
+	
+		// Create data file if it doesn't exist
+		int fd;
+		fd = open(AESD_SOCKET_DATA_FILE, O_RDWR | O_APPEND | O_CREAT, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+		if(fd == -1)
+		{
+			perror("socket data file open");
+			close(acceptedSocketfd);
+			freeaddrinfo(serverInfo);
+			close(serverSockfd);
```

Comment:
@suryakanteti think about how you could structure your program to do cleanup in one spot rather than duplicating the cleanup logic in multiple error handlers.